File: traditional_comm_py/traditional_comm/controller.py
from __future__ import annotations

import json
from pathlib import Path

from .runner import create_run_id, run_one


class TraditionalCommunicationController:
    def __init__(self, runs_dir: Path, transport):
        self.runs_dir = Path(runs_dir).resolve()
        self.transport = transport
        self.mode = "traditional"
        self.states: dict[str, dict] = {}

    def health_check(self) -> dict:
        return self.transport.health_check()

    def switch_mode(self, mode: str) -> dict:
        if mode not in {"traditional", "semantic"}:
            raise ValueError(f"unsupported mode: {mode}")
        self.mode = mode
        return {"mode": self.mode}

    def start_run(self, config: dict) -> dict:
        if self.mode != "traditional":
            raise RuntimeError("this controller executes traditional mode only")
        kind = config["kind"]
        input_path = Path(config["input_path"])
        run_id = create_run_id(kind)
        state = {
            "run_id": run_id,
            "mode": self.mode,
            "status": "running",
            "kind": kind,
            "input_path": str(input_path.resolve()),
        }
        self.states[run_id] = state
        try:
            result = run_one(
                kind,
                input_path,
                self.runs_dir,
                self.transport,
                config.get("task", {}),
                run_id=run_id,
            )
            state.update({"status": result["metrics"]["status"], "result": result})
        except Exception as exc:
            state.update({"status": "failed", "error": str(exc)})
            raise
        return {"run_id": run_id, "status": state["status"]}

    def get_status(self, run_id: str) -> dict:
        state = self.states[run_id]
        return {key: state.get(key) for key in ("run_id", "mode", "status", "kind", "input_path", "error")}

    def get_performance(self, run_id: str) -> dict:
        return self.states[run_id]["result"]["metrics"]

    def get_result(self, run_id: str) -> dict:
        result_dir = Path(self.states[run_id]["result"]["run_dir"])
        return json.loads((result_dir / "result.json").read_text(encoding="utf-8"))

    def stop_run(self, run_id: str) -> dict:
        state = self.states[run_id]
        if state["status"] == "running":
            state["status"] = "cancelled"
        return self.get_status(run_id)

File: traditional_comm/src/controller.js
const fs = require('node:fs');
const path = require('node:path');

const { createRunId, runOne } = require('./runner');

class TraditionalCommunicationController {
  constructor({ runsDir, transport }) {
    if (!runsDir) throw new Error('runsDir is required');
    if (!transport) throw new Error('transport is required');
    this.runsDir = path.resolve(runsDir);
    this.transport = transport;
    this.mode = 'traditional';
    this.states = new Map();
  }

  async healthCheck() {
    return this.transport.healthCheck();
  }

  switchMode(mode) {
    if (!['traditional', 'semantic'].includes(mode)) {
      throw new Error(`Unsupported mode: ${mode}`);
    }
    this.mode = mode;
    return { mode: this.mode };
  }

  async startRun({ kind, inputPath, task = {} }) {
    if (this.mode !== 'traditional') {
      throw new Error('This controller only executes traditional mode; semantic mode needs its own adapter');
    }
    const runId = createRunId(kind);
    const state = {
      run_id: runId,
      mode: this.mode,
      status: 'preparing',
      kind,
      input_path: path.resolve(inputPath),
    };
    this.states.set(runId, state);

    const promise = runOne({
      runId,
      kind,
      inputPath,
      runsDir: this.runsDir,
      transport: this.transport,
      task,
    }).then(result => {
      this.states.set(runId, { ...state, status: result.metrics.status, result });
      return result;
    }).catch(error => {
      this.states.set(runId, { ...state, status: 'failed', error: error.message });
      throw error;
    });
    state.status = 'running';
    state.promise = promise;
    return { run_id: runId, status: state.status };
  }

  async waitForRun(runId) {
    const state = this.states.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (state.promise) return state.promise;
    return state.result;
  }

  getStatus(runId) {
    const state = this.states.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    return {
      run_id: runId,
      mode: state.mode,
      status: state.status,
      kind: state.kind,
      input_path: state.input_path,
      error: state.error || null,
    };
  }

  getPerformance(runId) {
    const state = this.states.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (!state.result) return { run_id: runId, status: state.status, metrics: null };
    return state.result.metrics;
  }

  getResult(runId) {
    const state = this.states.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (!state.result) return { run_id: runId, status: state.status, result: null };
    const resultPath = path.join(state.result.runDir, 'result.json');
    return JSON.parse(fs.readFileSync(resultPath, 'utf8'));
  }

  stopRun(runId) {
    const state = this.states.get(runId);
    if (!state) throw new Error(`Unknown run_id: ${runId}`);
    if (state.status === 'running' || state.status === 'preparing') {
      state.status = 'cancelled';
    }
    return this.getStatus(runId);
  }
}

module.exports = { TraditionalCommunicationController };
